### DISCUSSION TITLE
Triggering changes to many source assets in single step

### DISCUSSION THREAD

**Discussion by the4thamigo-uk** [dagster-io/dagster#18854] [Q&A] [answered]
I read data from a third party s3 bucket which contains hourly files, and currently I have an hourly partitioned asset that materializes these files into our own data store.

Sometimes the third party updates  a file for a particular hour. I currently deal with this using a sensor by reading all the s3 keys   and removing the files with a timestamp which is less than or equal to the cursor and then launching runs for each new partition.

I would like to treat the upstream hourly files as source assets, as it seems this reflects what their role is in my pipeline, and in principle I can write an observation function to detect if there is a more recent file under the prefix for each hourly partition.

The problem with this approach, is that I would need to perform a boto s3 query for each and every partition in order to observe changes to the source asset. Instead, it would be much more efficient to query s3 once (perhaps using the sensor), to find all the modified hourly files, then update the data version for each modified hourly partition of the source asset. Is there a way I can achieve this?

EDIT: Ive just been reading about external assets, this maybe is the way forward, but still experimental.

**Comment by smackesey** [comment, marked as answer]
>The problem with this approach, is that I would need to perform a boto s3 query for each and every partition in order to observe changes to the source asset. Instead, it would be much more efficient to query s3 once (perhaps using the sensor), to find all the modified hourly files, then update the data version for each modified hourly partition of the source asset. Is there a way I can achieve this?

You can currently return a `DataVersionsByPartition` object from your observable source asset that will update data versions for all passed partitions:

```
@observable_source_asset(partitions_def=...)
def foo():
    query_result = ...
    # keys are partition keys, values are `DataVersion` objects
    return DataVersionsByPartition({
        "a": get_data_version_from_query_result(query_result, "a"),
        "b": get_data_version_from_query_result(query_result, "b"),
        ...
    })
```

But in the near future this functionality will be folded into external assets.

**Reply by the4thamigo-uk** [reply]
Is there a view on how soon external assets will go to full production? Im planning to go into production in Q1 2024, and not sure whether to go down the source asset or external asset route.




By the way thanks so much for responding to my query, it can be a bit of a struggle working through the docs and source code to figure out how to do things. However, it seems that if I were to use `DataVersionsByPartition` I would have to add an entry for each hourly file that is in S3. In most cases the data versions wont have changed, so this would presumably produce lots of needless `Asset Observations` unless I could filter out only the ones that have a data version different from the current data version of the partition... is there any way to optimize this?

**Reply by smackesey** [reply]
>Is there a view on how soon external assets will go to full production? Im planning to go into production in Q1 2024, and not sure whether to go down the source asset or external asset route.

We're looking at a mid-Feb full release right now, but the APIs should become available in January. I recommend just using observable source assets, we are going to automatically convert them into external assets under the hood anyway, and it probably wouldn't be much work to change to the new APIs when they are available.

>However, it seems that if I were to use DataVersionsByPartition I would have to add an entry for each hourly file that is in S3. In most cases the data versions wont have changed, so this would presumably produce lots of needless Asset Observations unless I could filter out only the ones that have a data version different from the current data version of the partition... is there any way to optimize this?

You'll only get new observations for the keys you include, so you can just omit the ones that you're not observing. If you instead want to collect a bunch, check which ones haven't changed, and only include those, I don't think we have a perfect API, but you could try `fetch_observations` to bulk fetch existing observations for the check: https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/instance/__init__.py#L1978-L2001

**Reply by the4thamigo-uk** [reply]
>  If you instead want to collect a bunch, check which ones haven't changed, and only include those

Yes this is the kind of thing I was looking for. Looks like it would be a bit more clunky to do this using source assets.

I have this kinda working for external assets, in that I keep track of the timestamp of the last file i read from S3, and then use this as a cursor in a sensor. I then know which files have been updated and I submit an `AssetMaterialization` for the external asset and partition.

Can I do a similar thing with the `SourceAsset`? I mean can I skip having an observation function for the `SourceAsset`, and simply submit my own `AssetObservation` events from a cursor for the source asset and partition?
